enable/improve constant propagation through varargs methods

jrevels opened the pull request:
This should allow constant propagation through varargs functions. I was just a keyboard monkey following @Keno's/@vtjnash's direction for this PR, so they can probably answer any questions better than I can.

EDIT: We've now added some other stuff that expands the scope here a bit (mainly so we don't forget some of these hacks), but this PR can be broken up into several others later if need be

jrevels committed:
enable/improve constant propagation through varargs methods

- Store varargs type information in the InferenceResult object, such that the info can be used during inference/optimization

- Hack in a more precise return type for getfield of a vararg tuple. Ideally, we would handle this by teaching inference to track the types of the individual fields of a Tuple, which would make this unnecessary, but until then, this hack is helpful.

- Spoof parents as well as children during recursion limiting, so that higher degree cycles are appropriately spoofed

- A broadcast test marked as broken is now no longer broken, presumably due to the optimizations in this commit

- Fix relationship between depth/mindepth in limit_type_size/is_derived_type. The relationship should have been inverse over the domain in which they overlap, but was not maintained consistently. An example of problematic case was:

t = Tuple{X,X} where X<:Tuple{Tuple{Int64,Vararg{Int64,N} where N},Tuple{Int64,Vararg{Int64,N} where N}}
c = Tuple{X,X} where X<:Tuple{Int64,Vararg{Int64,N} where N}

because is_derived_type was computing the depth of usage rather than the depth of definition. This change thus makes the depth/mindepth calculations more consistent, and causes the limiting heuristic to return strictly wider types than it did before.

- Move the optimizer's "varargs types to tuple type" rewrite to after cache lookup.Inference is populating the InferenceResult cache using the varargs form, so the optimizer needs to do the lookup before writing the atypes in order to avoid cache misses.

Co-authored-by: Jameson Nash <vtjnash@users.noreply.github.com>
Co-authored-by: Keno Fischer <keno@alumni.harvard.edu>

vtjnash commented on base/compiler/inferenceresult.jl:
```diff
@@ -31,7 +32,32 @@ function get_argtypes(result::InferenceResult)
+            result.vargs = vararg_type_vec
         end
         args[nargs] = vararg_type
         nargs -= 1
```
Ah, wait, no. I think tuple_tfunc took care of this

iblis17 commented:
BSD CI got unkillable process from this PR, and it ate 100% CPU on a core.
Here is the bt from LLDB, hope this help.
https://gist.github.com/iblis17/df720508da46677c3f5d73318a136362

vtjnash commented on base/compiler/inferenceresult.jl:
```diff
@@ -31,7 +32,32 @@ function get_argtypes(result::InferenceResult)
+                if isvarargtype(va)
+                    # assumes that we should never see Vararg{T, x}, where x is a constant (should be guaranteed by construction)
+                    va = rewrap_unionall(va, linfo.specTypes)
+                    vararg_type_vec = Any[va]
```
actually, looks like the inference code will handle this, so this is OK

jrevels replied:
So to make sure I follow, you're saying that we can't do anything if we don't actually have the types of the "trailing" arguments in `atypes`, so this should just be:

```julia
if nargs > laty
    vararg_type_vec = Any[]
    vararg_type = Tuple{}
else
    ⋮ # the stuff that's already here
end
```

jrevels commented on base/compiler/inferenceresult.jl:
```diff
@@ -31,7 +32,32 @@ function get_argtypes(result::InferenceResult)
+                vararg_type_vec = Any[rewrap_unionall(p, linfo.specTypes) for p in atypes[nargs:laty]]
+                vararg_type = tuple_tfunc(Tuple{vararg_type_vec...})
+                for i in 1:length(vararg_type_vec)
+                    atyp = vararg_type_vec[i]
```
(Just talked to Jameson in person, turns out this is actually okay)

jrevels commented on base/compiler/inferenceresult.jl:
```diff
@@ -75,22 +101,12 @@ end
-        if cache_code.linfo === code && length(cache_args) >= nargs
+        if cache_code.linfo === code && length(cache_args) == length(argtypes)
             cache_match = true
-            # verify that the trailing args (va) aren't Const
```
Verifying this correctly would essentially require fully recomputing the `vararg_type_vec` part of `get_argtypes`, right? 

Should I actually do that, or is there a more lightweight check I should be doing?

jrevels commented:
Looks like this passes all tests except for `compiler/compiler.jl`, where it hangs.

jrevels commented:
Okay, I've tracked down the offending test here: https://github.com/JuliaLang/julia/blob/1c5ed7082516c8a8c259907ceef908ba0c9d2107/test/compiler/compiler.jl#L1248

Interrupting that test while it hangs yields:

```
^CWARNING: Force throwing a SIGINT
Internal error: encountered unexpected error in runtime:
InterruptException()
jl_egal at /Users/jarrettrevels/data/repos/julia-dev/src/builtins.c:171
is_derived_type at ./compiler/typelimits.jl:66
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:74
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
⋮
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:74
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:74
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:74
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:55
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type at ./compiler/typelimits.jl:74
jfptr_is_derived_type_2599 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
is_derived_type_from_any at ./compiler/typelimits.jl:94
jfptr_is_derived_type_from_any_2596 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
type_more_complex at ./compiler/typelimits.jl:228
jfptr_type_more_complex_2588 at /Users/jarrettrevels/data/repos/julia-dev/usr/lib/julia/sys.dylib (unknown line)
⋮
```

jrevels commented:
Woohoo, tests here now pass locally.

jrevels commented:
Only thing left here now are adding tests and investigating the potential cache problems pointed out at https://github.com/jrevels/Cassette.jl/pull/41#issuecomment-383353608

jrevels commented:
Okay, tests added and caching problem fixed (thanks to @vtjnash). I rebased and squashed so that this doesn't leave a bunch of intermediate broken commits. Only thing left now is to make sure CI passes and then this is good to go.

jrevels commented:
From running tests locally, it looks like everything passes except a number of tests in `SparseArrays/test/higherorderfns.jl` - including the tests that are marked `@test_broken` on master which this PR had previously fixed. Looks like we'll have to get those passing again before merging this...

mbauman commented:
Good luck. I've lost many hours to failures in there.  Typically it's been due to a failure to inline.  Check `@code_typed` and search for `invoke`s.  There should be very few and none in the non-error path.

jrevels commented:
So, on master (see https://gist.github.com/jrevels/20f5e1d7ac64c0b51099960ded3e4687 for test setup):

```julia
julia> @test (@allocated broadcast!(+, Q, X, Y, Z)) == 0
Test Passed

julia> @test (@allocated broadcast!(*, Q, X, Y, Z)) == 0
Test Passed

julia> @test (@allocated broadcast!(f, Q, X, Y, Z)) == 0
Test Failed at REPL[10]:1
  Expression: #= REPL[10]:1 =# @allocated(broadcast!(f, Q, X, Y, Z)) == 0
   Evaluated: 16 == 0
ERROR: There was an error during testing
```

On this PR currently:

```julia
julia> @test (@allocated broadcast!(+, Q, X, Y, Z)) == 0
Test Failed at REPL[8]:1
  Expression: #= REPL[8]:1 =# @allocated(broadcast!(+, Q, X, Y, Z)) == 0
   Evaluated: 6688 == 0
ERROR: There was an error during testing

julia> @test (@allocated broadcast!(*, Q, X, Y, Z)) == 0
Test Failed at REPL[9]:1
  Expression: #= REPL[9]:1 =# @allocated(broadcast!(*, Q, X, Y, Z)) == 0
   Evaluated: 6688 == 0
ERROR: There was an error during testing

julia> @test (@allocated broadcast!(f, Q, X, Y, Z)) == 0
Test Failed at REPL[10]:1
  Expression: #= REPL[10]:1 =# @allocated(broadcast!(f, Q, X, Y, Z)) == 0
   Evaluated: 1744 == 0
ERROR: There was an error during testing
```

This PR, but reverting the [change to inlineable that reduced InferenceResult cache misses](https://github.com/JuliaLang/julia/pull/26826/files#diff-d4c4d6c8ee96fc79b74ac240f5b84ebbL1302):

```julia
julia> @test (@allocated broadcast!(+, Q, X, Y, Z)) == 0
Test Failed at REPL[8]:1
  Expression: #= REPL[8]:1 =# @allocated(broadcast!(+, Q, X, Y, Z)) == 0
   Evaluated: 96 == 0
ERROR: There was an error during testing

julia> @test (@allocated broadcast!(*, Q, X, Y, Z)) == 0
Test Failed at REPL[9]:1
  Expression: #= REPL[9]:1 =# @allocated(broadcast!(*, Q, X, Y, Z)) == 0
   Evaluated: 96 == 0
ERROR: There was an error during testing

julia> @test (@allocated broadcast!(f, Q, X, Y, Z)) == 0
Test Failed at REPL[10]:1
  Expression: #= REPL[10]:1 =# @allocated(broadcast!(f, Q, X, Y, Z)) == 0
   Evaluated: 96 == 0
ERROR: There was an error during testing
```

Before I rebased, all these tests were passing...argh...

jrevels commented:
D'oh, I was dumb and didn't rebuild the sysimg. It turns out the culprit is entirely the [change to inlineable that reduced InferenceResult cache misses](https://github.com/JuliaLang/julia/pull/26826/files#diff-d4c4d6c8ee96fc79b74ac240f5b84ebbL1302), which makes sense.

Here's this PR on the MWE tests with the offending change reverted:

```julia
julia> @test (@allocated broadcast!(+, Q, X, Y, Z)) == 0
Test Passed

julia> @test (@allocated broadcast!(*, Q, X, Y, Z)) == 0
Test Passed

julia> @test (@allocated broadcast!(f, Q, X, Y, Z)) == 0
Test Passed
```

So I just need to figure out how to fix those cache misses without affecting inlining...

jrevels commented:
Okay, I just pushed a rewrite of the cache lookup which passes both the new compiler tests and the MWE above. 

@vtjnash points out that the previous cache lookup fix also fixed another problem that is currently on master: the [`argexprs` passed to `inline_as_constant(linfo.inferred_const, argexprs, sv, invoke_data)`](https://github.com/JuliaLang/julia/pull/26826/files?utf8=%E2%9C%93&diff=unified#diff-d4c4d6c8ee96fc79b74ac240f5b84ebbR1331) should be pre-varargs-to-tuple-type rewrite (what is currently called `argexprs0` in the linked snippet). However, this fix apparently exposed a different issue in the optimizer, which is what caused the sparse broadcast test breakage.

I'm not going to consider fixing this issue necessary for merging this PR, since the issue already exists on master, but it might be worth looking into later.

jrevels commented:
Okay, tests all pass locally now. This PR is basically done, so I'm going to merge once CI turns up green unless there are any objections.

KristofferC commented:
Worth running Nanosoldier?

jrevels commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/26826#issuecomment-386418370) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/7483ee4e1eb2e6ee03961ee051b2999ccfecf938/daf35b2_vs_fe99c4a/report.md). cc @ararslan

JeffBezanson commented on base/compiler/inferenceresult.jl:
```diff
@@ -5,6 +5,7 @@ const EMPTY_VECTOR = Vector{Any}()
 mutable struct InferenceResult
     linfo::MethodInstance
     args::Vector{Any}
+    vargs::Vector{Any}
```
This field could use an explanatory comment.

jrevels commented:
Added the requested comment. Investigated the benchmark regressions, none seemed significant. Even the 3.65x one showed no regression locally, and had the same `@code_typed` output here vs. master. I guess nanosoldier is really noisy these days...

CI was totally green (besides unrelated Windows timeout) before I added the requested comment, so I'll go ahead and merge (I did rebuild and run tests locally just be absolutely sure, but of course adding a comment didn't break anything).
